### DISCUSSION TITLE
[Langium Grammar] How best to implement section headings?

### DISCUSSION THREAD

**Discussion by ym-han** [eclipse-langium/langium#1592] [Q&A] [answered]
What's the best way to implement optional section headings (much like Markdown's) like

```
SECTION 1: Introduction

<... stuff ...>

SECTION 2

<... stuff ...>
```

where there aren't super-explicit closing delimiters?

I see from https://github.com/eclipse-langium/langium/discussions/153 that one way to do it is to use `EOL` --- is that the best way?

Thanks very much in advance!

**Comment by msujew** [comment, marked as answer]
Hey @ym-han,

sorry for the late response. You can use a parser rule like this to identify the end of section headings:

```antlr
Section: 'SECTION' content=TEXT (EOL+ | EOF); // EOF is a builtin keyword that matches the end of a file

terminal EOL: /\r?\n/;
terminal TEXT: [^\n\r]+;
```

A general word of warning: Markdown like languages are very difficult to parse using LL based parsers (like Langium). Parsers for those languages are usually handwritten due to very specific behavior wrt to stuff like line endings, ambiguity detection and error recovery. It can work, but it's often a lot more effort than necessary.

**Reply by ym-han** [reply]
Gotcha, thank you! Sounds like there's no way to do it without using `EOL` then!
